rfc-149-switch-to-per-page-asset-loading.md

```diff
@@ -0,0 +1,208 @@
+
+### Non-rendering applications that use the components gem
+
+There are applications that use the components gem, but don't render public pages on GOV.UK - for example, publishing applications and the Govspeak gem. These may have a different set up with different ways of rendering pages.
+
+These changes should be optional - even if this is the default behaviour, it should be able to be turned off at a component by component level and at an application level. This could be done with a helper method that checks both if the component has already been used, and if adding the component assets are allowed:
+
+```ruby
+# component_asset_helper.rb - pseudo code example:
+
+def component_assets_allowed (component_name)
+  # check:
+  #  - if the component has already had the assets added to the page
+  # - if the component assets loading method is allowed at an application level
+  # - if the individual component has used an option to turn individual assets on or off
+  @components_in_use.include?(component_name) || ENV.component_asset_loading !== "bulk" || use_individual_assets === true
+end
+```
+
+## Other solutions that have been considered and rejected
+
+* Put all components used across GOV.UK in one stylesheet and one JavaScript file - this increases page weight and doesn't solve cache invalidation
+* Serve one stylesheet and one JavaScript file for all of GOV.UK - this increases page weight and doesn't solve cache invalidation
+* Serve one stylesheet and one JavaScript file per template - this doesn't solve cache invalidation when one component changes
+* Serve all component assets used in an application in individual files - this increases page weight for single page visitors, though often decreased it for people who visited more than one page
+
+## Implementation process
+
+Serving assets individually doesn't rely on the shared asset folder proposed in RFC 91 - so the lack of a shared folder is not a blocker.
+
+* implement the individual serving of assets
+   1. Stylesheets
+   1. JavaScript, as this will require more investigation and relies on external factors
+* once a shared folder is available, make all applications load assets from the shared folder
+
+### Implement the individual serving of assets
+
+Splitting this into two stages makes it more manageable and will allow improvements to be in stages.
+
+The stages themselves can be split into app-by-app upgrades - once an app is upgraded the pages it renders will see performance improvements without relying on any other app also being upgraded. This is a similar approach to the work that was done to use the public layout component, so we know it can work - even if it takes time.
```

Comment:
Hmm, I'm not sure there is anything written down about the roll out of the public layout component. There are individual pull requests to update each application, so that could show the time between the first and last app to be upgraded - do you think that would be useful?

Reply:
A link to one of the PRs sounds good 👍 i.e. `This is a similar approach to the work that was done to use the public layout component ([example](#)), so we know it can work`.

Reply:
I've added examples from near the start of the work (collections) and the end (whitehall of course) in https://github.com/alphagov/govuk-rfcs/pull/152/commits/07022a4f09ea4b3295ccd0c5273a5e31d5ed57e0 - is this okay?

Reply:
Brilliant, thanks 👍